### DISCUSSION TITLE
RFC: Override page component

### DISCUSSION THREAD

**Discussion by migueloller** [makeswift/makeswift#25] [Ideas]
# Background
We've heard feedback from @cfofiu that it would be nice if a developer could set up a page in Makeswift in a way such that the Makeswift user can only edit specific sections of the page. This would be useful to lock things down or incrementally adopt Makeswift for an existing page.

# Proposal
When building a Makeswift page, the root element is automatically created and looks roughly something like this:

```json
{
  "type": "<page_component_type>",
  "key": "<key>",
  "props": {}
}
```

While we need a root element to get the page started, the `Page` component that renders that element would be overridable in the Makeswift runtime. It could look something like this:

```typescript
import { ReactRuntime } from '@makeswift/runtime/react'
import { Props as Controls } from '@makeswift/runtime/prop-controllers'

import MyPage from './path/to/my/page-component'

ReactRuntime.registerPageComponent(MyPage, {
  props: {
    children: Controls.Slot() // NOTE: The `Slot` control is still being worked on.
  }
})
```

By allowing the page component to be overwritten, it becomes possible to create a custom page component that exposes certain sections. i.e.:

```typescript
import { Header, Footer } from '@/components'
import { ReactRuntime } from '@makeswift/runtime/react'

function MyPage({ top, bottom }) {
  return (
    <main>
      {/* <Header /> is a code component that's not registered with Makeswift. */}
      <Header />
      <section id="top">{top}</section>
      <section id="middle">{/* custom code goes here */}</section>
      <section id="bottom">{bottom}</section>
      {/* <Footer /> is a code component that's not registered with Makeswift. */}
      <Footer />
    </main>
  )
}

ReactRuntime.registerComponent(MyPage, {
  props: {
    top: Controls.Slot(),
    bottom: Controls.Slot(),
  }
})
```

The example above would allow the `top` and `bottom` sections to be editable in the Makeswift builder but everything else would not be editable since it's just plain React, not registered with Makeswift in any way.

# Backwards Compatibility
This is a new feature that wouldn't affect existing APIs so it should be backwards compatible.

# Release Timeline
This is in the idea stage and requires the `Slot` control to be released before it can be useful. There's not a specific timeline on this at the time of writing.

# Feedback
We would love to hear your feedback on this idea and whether it works for your team and why.

**Comment by migueloller** [comment]
Note that at the moment there is a "hack" to get similar behavior where there are parts of the page that aren't editable. It looks something like this:

```typescript
import { RuntimeProvider, Document as DocumentComponent } from '@makeswift/runtime/react'

export default function Page({ pageId, rootElement, makeswiftApiEndpoint }: PageProps) {
  return (
    <RuntimeProvider
      defaultRootElements={new Map([[pageId, rootElement]])}
      makeswiftApiEndpoint={makeswiftApiEndpoint}
    >
      <main>
        {/* <Header /> is a code component that's not registered with Makeswift. */}
        <Header />
        <DocumentComponent documentKey={pageId} />
        {/* <Footer /> is a code component that's not registered with Makeswift. */}
        <Footer />
      </main>
    </RuntimeProvider>
  )
}
```

You would then use this page component instead of using the one exported by `@makeswift/runtime/next`. Note that this is just "embedding" the Makeswift page document within the `<main>` element and so you can only have one "slot" instead of multiple ones via props. Hopefully it also goes without saying that while we're in prerelease these APIs are changing very quickly and this "hack" might not work forever.

**Comment by apledger** [comment]
What if I want to only have one page use the custom page? I would still want the custom page to be available in the Makeswift UI, but other pages to use the base Makeswift page.

**Reply by migueloller** [reply]
The `ReactRuntime.registerPageComponent` call can be called on specific pages so you would only call that in the pages that you want to override. As far as the specifics of how you separate a page out of the catch-all (i.e., `[[...path]].js`) you can just make a separate Next.js page and render the Makeswift components there. We should add some examples on how to do that and also draft some new RFCs for registering pre-existing pages with Makeswift (i.e., instead of creating the page in the Makeswift UI, the page is registered from the host and it appears in the sidebar).
